### PR TITLE
#34893: Run toolkit manager bootstrap logic in a background thread. (merges into develop/plugin)

### PR DESCRIPTION
**Pull request by manneohrstrom**
Same as #19 but merging into `develop/plugin` instead of master.

### CODE REVIEW

File: plugins/basic/python/tk_maya_basic/plugin_engine.py [hunk 1]
```diff
@@ -53,25 +52,29 @@ def bootstrap(sg_user):
...
+    # followed by launching the engine synchronously in the main application thread.
+    toolkit_mgr.async_bootstrap = True
+
     # Before bootstrapping the engine for the first time around,
     # the toolkit manager may swap the toolkit core to its latest version.
-    toolkit_mgr.bootstrap_engine(manifest.engine_name)
-
-    # Re-import the toolkit core to ensure usage of a swapped in version.
-    import sgtk
+    toolkit_mgr.bootstrap_engine(manifest.engine_name,
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_engine.py]: I suggest using this indentation instead:

```python
toolkit_mgr.bootstrap_engine(
    manifest.engine_name,
    progress_callback=progress_callback,
    completed_callback=completed_callback,
    failed_callback=failed_callback
)
```
We are trying to move a lot of our code base across to this style at the moment!

File: plugins/basic/python/tk_maya_basic/plugin_engine.py [hunk 2]
```diff
@@ -53,25 +52,29 @@ def bootstrap(sg_user):
...
 
 def shutdown():
     """
     Shuts down the running engine.
     """
 
+    # Re-import the toolkit core to ensure usage of a swapped in version.
+    import sgtk
+
     logger = sgtk.LogManager.get_logger(PLUGIN_PACKAGE_NAME)
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_engine.py]: What's the benefit with this constant? I wonder if it's perhaps overkill? Could we just do `logger = sgtk.LogManager.get_logger(__name__)` (this also matches the [docs](http://developer.shotgunsoftware.com/tk-core/utils.html#logging-from-scripts-and-other-external-locations) )

File: plugins/basic/python/tk_maya_basic/plugin_engine.py [hunk 3]
```diff
@@ -53,25 +52,29 @@ def bootstrap(sg_user):
...
 
-    logger.info("Starting the %s engine." % manifest.engine_name)
-
     # Remove the custom logging handler now that the engine will take over logging.
     sgtk.LogManager().root_logger.removeHandler(plugin_logging_handler)
 
-    # Ladies and Gentlemen, start your engines!
+    # Bootstrap a toolkit instance asynchronously in a background thread,
+    # followed by launching the engine synchronously in the main application thread.
+    toolkit_mgr.async_bootstrap = True
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_engine.py]: wohoo!

File: plugins/basic/python/tk_maya_basic/plugin_logic.py [hunk 4]
```diff
@@ -43,6 +44,9 @@
 # Use a custom logging handler to display messages in Maya script editor.
 standalone_logger.addHandler(plugin_logging.PluginLoggingHandler(manifest.name))
 
+# Main progress bar used when bootstrapping the toolkit and engine.
+main_progress_bar = None
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: do we need to keep a global around for this? Maya already does this for us? In my experience, I tend to try as few 'pointers' into maya as possible. Would it make sense to do 

```
def get_maya_main_progress_bar():
    return pm.ui.MainProgressBar(minValue=0, maxValue=100, interruptable=False)
```

What happens if the plugin is downloading stuff from the app store and the user keeps working, doing other operations which also utilises the main progress bar? For this reason, I think it may be safer if the bootstrap is getting the progress bar every time.

File: plugins/basic/python/tk_maya_basic/plugin_logic.py [hunk 5]
```diff
@@ -75,60 +79,134 @@ def login_user():
...
+                        from 1 to ``sgtk.bootstrap.ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+    :param message: Progress message to report.
+    :param current_index: Optional current item number being looped over.
+    :param maximum_index: Optional maximum item number being looped over.
+    """
+
+    if maximum_index and maximum_index > 1:
+        message = "%s (%s of %s)" % (message, current_index+1, maximum_index+1)
+
+    message = "Bootstrapping %s: %s" % (manifest.engine_name, message)
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: I guess i need to see what this looks like in maya, but wondering if this should say  **Shotgun:** *Downloading v1.2.3 of tk-loader...*. I don't think artists understand what 'bootstrapping' means.

File: plugins/basic/python/tk_maya_basic/plugin_logic.py [hunk 6]
```diff
@@ -75,60 +79,134 @@ def login_user():
...
+
+    This function is executed in the main thread by the main event loop.
+
+    :param engine: Launched :class:`sgtk.platform.Engine` instance.
+    """
 
-        try:
-            pm.waitCursor(state=True)
+    # Needed global to re-import the toolkit core.
+    global sgtk
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: why is this needed?

File: plugins/basic/python/tk_maya_basic/plugin_logic.py [hunk 7]
```diff
@@ -75,60 +79,134 @@ def login_user():
...
+    # Get rid of the main progress bar.
+    main_progress_bar.endProgress()
+    main_progress_bar = None
 
-        # Add a logout menu item to the engine context menu.
-        sgtk.platform.current_engine().register_command(ITEM_LABEL_LOGOUT,
-                                                        logout_user,
-                                                        {"type": "context_menu"})
+    # Report the encountered exception.
+    standalone_logger.info("Bootstrapping %s failed: %s" % (manifest.engine_name, exception))
```
Comment (manneohrstrom) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: seems like this should be an error rather than just info.

Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: Done.

File: plugins/basic/python/tk_maya_basic/plugin_logic.py (hunk 6 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: The `global sgtk` is needed in order for the local `import sgtk` to update `sgtk` at the module level, since `sgtk` is used in other functions.

Comment (josh-t) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: I think that once the bootstrap is complete, the old `sgtk` is cleared from `sys.modules` so the next subsequent import should get the new one. I could easily be wrong or misunderstanding this case though. Maybe this is what @manneohrstrom was expecting too?

File: plugins/basic/python/tk_maya_basic/plugin_logic.py (hunk 5 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: This is the list of messages that are displayed aside the main progress bar on Maya Help line as the bootstrap progresses:

> Bootstrapping tk-maya: Resolving project...
> Bootstrapping tk-maya: Resolving configuration...
> Bootstrapping tk-maya: Updating configuration...
> Bootstrapping tk-maya: Starting up Toolkit...
> Bootstrapping tk-maya: Downloading tk-maya v0.5.10 (1 of 5)...
> Bootstrapping tk-maya: Downloading tk-multi-shotgunpanel v1.2.6 (2 of 5)...
> Bootstrapping tk-maya: Downloading tk-multi-loader2 v1.11.2 (3 of 5)...
> Bootstrapping tk-maya: Downloading tk-framework-shotgunutils v4.3.1 (4 of 5)...
> Bootstrapping tk-maya: Downloading tk-framework-qtwidgets v2.2.1 (5 of 5)...
> Bootstrapping tk-maya: Resolving context...
> Bootstrapping tk-maya: Launching Engine...

Updating to say "Initializing Shotgun:".

File: plugins/basic/python/tk_maya_basic/plugin_logic.py (hunk 6 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: This is the case, but in order to not have to do an `import sgtk` in each function of the module, using the `global` makes module `sgtk` _imported name variable_ updates.

Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: From https://docs.python.org/2/reference/simple_stmts.html#the-import-statement:

> The first form of import statement binds the module name in the local namespace to the module object, ...

In other words, without the `global`, the `sgtk` name binding at the `plugin_logic` level would not be updated.

File: plugins/basic/python/tk_maya_basic/plugin_logic.py (hunk 4 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_logic.py]: After some experimentation, now using a variation of this approach, with no more global since there is only one main progress bar.

In between plug-in progress bar updates, other user interactions may overwrite the progress message, and possibly the progress bar. At the next plug-in update, the new code makes sure the progress bar uses the proper configuration and updates the progress value and message.

File: plugins/basic/python/tk_maya_basic/plugin_engine.py (hunk 2 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_engine.py]: Using `__file__` would display information/warning/error messages like this one to the user in Maya script editor:

> // Shotgun: plugin_engine: Stopping the tk-maya engine. //

Using `PLUGIN_PACKAGE_NAME` replaces `plugin_engine` by the more descriptive plug-in package name, for the time being `tk_maya_basic`.

Ideally, tk-core engine logging and tk-maya should allow setting the message prefix rather than forcing the not-very-user-friendly predefined format `Shotgun: %(basename)s: `. This would allow using as prefix the plug-in name defined in the `info.yml` file.

File: plugins/basic/python/tk_maya_basic/plugin_engine.py (hunk 1 above)
Comment (fpainchaud) [inline plugins/basic/python/tk_maya_basic/plugin_engine.py]: OK. Done.
